client/finality-grandpa/src/lib.rs

```diff
@@ -541,6 +541,24 @@ where
 	)
 }
 
+/// A descriptor for an authority set hard fork. These are authority set changes
+/// that are not signalled by the runtime and instead are defined off-chain
+/// (hence the hard fork).
+pub struct AuthoritySetHardFork<Block: BlockT> {
+	/// The new authority set id.
+	pub set_id: SetId,
+	/// The block hash and number at which the hard fork should be applied.
+	pub block: (Block::Hash, NumberFor<Block>),
+	/// The authorities in the new set.
+	pub authorities: AuthorityList,
+	/// The latest block number that was finalized before this authority set
+	/// hard fork. When defined, the authority set change will be forced, i.e.
+	/// the node won't wait for the block above to be finalized before enacting
+	/// the change, and the given finalized number will be used as a base for
+	/// voting.
+	pub last_finalized: Option<NumberFor<Block>>,
```

Comment:
Sorry again me :P 

Does this mean that `block` above should be `last_finalized + 1`?

Reply:
It doesn't necessarily need to be `+ 1`, but it needs to be a block that is higher than `last_finalized` yes.

Reply:
I'm completely confused now. 

Let's say we set `last_finalized` to `1` and `block` is `3`. So, the set will be enacted at `3`, but it will already be active at `1`?

Reply:
In that scenario the change gets enacted when block #3 is imported, and the GRANDPA voter will act as if block #1 is finalized, i.e. this is the base it will start voting on.

(I think you are forgetting that GRANDPA needs to distinguish between best imported and best finalized, and best finalized is always lower or equal to best imported)

Reply:
> In that scenario the change gets enacted when block #3 is imported, and the GRANDPA voter will act as if block #1 is finalized, i.e. this is the base it will start voting on.

With the new set?